src/mgr/DaemonServer.cc

```diff
@@ -1643,16 +1645,17 @@ bool DaemonServer::_handle_command(
 	    continue;
 	  }
 	  touched_pgs++;
-	  if (!(q.second.state & PG_STATE_ACTIVE) ||
-	      (q.second.state & PG_STATE_DEGRADED)) {
+	  if (!(q.second.state & PG_STATE_ACTIVE)) {
 	    ++dangerous_pgs;
 	    continue;
 	  }
 	  const pg_pool_t *pi = osdmap.get_pg_pool(q.first.pool());
 	  if (!pi) {
 	    ++dangerous_pgs; // pool is creating or deleting
 	  } else {
-	    if (pg_acting.size() < pi->min_size) {
+	    if (((q.second.state & PG_STATE_DEGRADED) &&
+                  !avail_no_missing.size()) ||
+		pg_acting.size() < pi->min_size) {
```

Comment:
It seems to that if the pg is DEGRADED, then pg_acting == avail_no_missing.  So avail_no_missing has no extra information.

Reply:
Since !avail_no_missing.size() == (pg_acting.size() == 0), then it has no change whether dangerous_pgs is incremented since 0 should always be < min_size.

Reply:
@dzafman Um...It seems that an OSD can be a member of pg_acting even if it has missing objects, so I think when the pg is DEGRADED, pg_acting may not be the same as avail_no_missing, right? And for the same reason, pg_acting may not be necessarily empty even if no OSD has all objects, and I think that's why we have a concept of "unfound objects". Am I right?:-)

Reply:
When a pg is degraded the _pg_acting_ is based on _avail_no_missing_.  If not, pg_acting comes from _acting_.  In any case the size of _pg_acting_ must be >= _min_size_.  Unfound is for objects that aren't stored on any OSD.

Now that think about it, I'm not clear how a pg with pg_acting.size() == min_size wouldn't be dangerous because stopping an OSD could then cause a PG to drop below min_size.  Doesn't "available" imply writable.